homeassistant/components/whirlpool/sensor.py

```diff
@@ -0,0 +1,180 @@
+"""The Washer/Dryer Sensor for Whirlpool Appliances."""
+from __future__ import annotations
+
+from collections.abc import Callable
+from dataclasses import dataclass
+import logging
+
+from whirlpool.washerdryer import MachineState, WasherDryer
+
+from homeassistant.components.sensor import (
+    SensorDeviceClass,
+    SensorEntity,
+    SensorEntityDescription,
+    SensorStateClass,
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import STATE_UNKNOWN
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.entity import DeviceInfo
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.helpers.typing import StateType
+
+from . import WhirlpoolData
+from .const import (
+    CYCLE_FUNC,
+    DOMAIN,
+    ICON_D,
+    ICON_W,
+    MACHINE_STATE,
+    TANK_FILL,
+    WASHER_STATE,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+def washer_state(washer: WasherDryer) -> str | None:
+    """Determine correct states for a washer."""
+
+    machine_state = washer.get_machine_state()
+    machine_cycle = None
+    for count, func in zip(range(len(CYCLE_FUNC)), CYCLE_FUNC):
+        if func(washer):
+            machine_cycle = WASHER_STATE.get(count)
+            break
+
+    if washer.get_attribute("Cavity_OpStatusDoorOpen") == "1":
+        machine_cycle = "Door open"
+        return machine_cycle
+
+    if machine_state == MachineState.RunningMainCycle and machine_cycle:
+        return machine_cycle
+
+    return MACHINE_STATE.get(machine_state, STATE_UNKNOWN)
+
+
+@dataclass
+class WhirlpoolSensorEntityDescriptionMixin:
+    """Mixin for required keys."""
+
+    value_fn: Callable
+
+
+@dataclass
+class WhirlpoolSensorEntityDescription(
+    SensorEntityDescription, WhirlpoolSensorEntityDescriptionMixin
+):
+    """Describes Whirlpool Washer sensor entity."""
+
+
+SENSORS: tuple[WhirlpoolSensorEntityDescription, ...] = (
+    WhirlpoolSensorEntityDescription(
+        key="state",
+        name="state",
+        entity_registry_enabled_default=True,
+        icon=ICON_W,
+        has_entity_name=True,
+        value_fn=washer_state,
+    ),
+    WhirlpoolSensorEntityDescription(
+        key="timeremaining",
+        name="time remaining",
+        device_class=SensorDeviceClass.DURATION,
```

Comment:
From https://developers.home-assistant.io/docs/core/entity/sensor?_highlight=duration#available-device-classes for `DURATION` says `Time period. Should not update only due to time passing. The device or service needs to give a new data point to update.`

My impression is this is updating as the cycle is counting down (anti-pattern described there), and the reason is to avoid unnecessary updates from the state machine just counting down time.  The idea instead would be to set a target time of when the cycle will finish instead, so it doesn't have to change once set for a timed cycle.

Reply:
This sensor is updated directly from the API via a JSON message.  It is not really time based as the washer and/or dryer take longer than originally estimated as the machine progresses (clothes don't dry as fast as anticipated, or the washer needs to re balance, water fill takes longer, etc).  Is it still "wrong" for this application?

Reply:
My impression that the estimated end time accuracy is not a factor in this (e.g. accuracy plays a role whether or not this is a specific timestamp or a time delta)

My impression is that the common pattern of having an estimated completion time is common across all integrations like sprinklers, battery chargers, etc.  and there are two factors in how this is updated:
- Estimates based on the current state of its view of the workload
- As time passes

And this is not saying that you can't still refine the estimated time as more is learned.  Updates for the first reason are fine, but updates for the second reason are a problem:  If it updates too rarely, then it seems like it's going to be stale and not reflect time passing so you'd want it to be more accurate.  If it's too fast to accurately reflect time passing then its straining the state machine.

It's OK if the machine regularly updates the new estimated time as it learns more.

Reply:
I probably misunderstand this one.  
Here's what I know about the messages: The device (washer or dryer) sends a message (based on the backend system configuration) updating this value.  It tends to send an update every minute, but sometimes a bit faster or slower. There is nothing I can do to "stop" it from sending the message as the API subscribes to an appliance and it (whirlpool backend) starts sending everything about the appliance.  
Should this be something other than DURATION for this behavior?

Reply:
I think I understand what it's doing: This is not recomputed any time the entity is asked for a new value. I see the value is in seconds and so this also means it is going to be incorrect/inaccurate between updates as well.

This is not my area of expertise in the system, but I don't think that changes much: it still sounds like from the perspective of the core state machine that using a timestamp is still the best practice e.g. like https://github.com/home-assistant/core/blob/fba13dcc907b85f84a2a3d60f3298858439793ca/homeassistant/components/rainmachine/sensor.py#L205 -- that example uses `SensorDeviceClass.TIMESTAMP` and `EntityCategory.DIAGNOSTIC`.

Reply:
Thanks for the example - it all makes a lot more sense now :)
I've updated the sensor to timestamp (and the complications that came with it)